# Análisis y Procesamiento de robots.txt y sitemaps

### Práctica 1

Mediante su browser, lea los archivos robots.txt de los siguientes sitios, analice las directivas que allí encuentre y comente con el mayor detalle posible.


https://www.cicese.edu.mx/robots.txt

La directiva de este sitio es simple ya que solo tiene disallow para dos tipos de bots, Semrush, que es un bot que se enfoca en la recopilación de datos para el análisis de marketing digital, y variaciones de este mismo bot con fines parecidos. El otro bot que bloquea es BLEXBot, el cual se enfoca en la recopilación de estructuras de URLs y interenlazado de sitios web, con esta directiva, la página indica que esos bots no deben rastrear ni indexar ninguna sección del sitio.


User-agent: SemrushBot
Disallow: /
User-agent: SemrushBot-SA
Disallow: /
User-agent: SemrushBot-BM
Disallow: / 
User-agent: SemrushBot-CT
Disallow: /
User-agent: SemrushBot-SWA
Disallow: / 
User-agent: SemrushBot-SI
Disallow: / 
User-agent: SemrushBot-BA
Disallow: / 
User-agent: BLEXBot
Disallow: /




https://www.uabc.mx/robots.txt

El robots de este sitio tiene una directiva que aplica por igual a todos los agentes, lo cual se indica con User-agent: *, esta directiva bloquea el directorio de administración de wordpress, exceptuando un archivo (Allow: /wp-admin/admin-ajax.php). De igual manera, también bloquea /wp-content/uploads/wpforms/, que es el directorio donde se almacenan archivos de WPForms.
Este robots, también incluye el URL para el sitemap



User-agent: *
Disallow: /wp-admin/
Allow: /wp-admin/admin-ajax.php
Disallow: /wp-content/uploads/wpforms/
Sitemap: https://www.uabc.mx/wp-sitemap.xml

https://www.unam.mx/robots.txt

Este sitio contiene un robots mucho más detallado, el cual de inicio indica que su directiva se aplica a todos los agentes por igual y que el delay para hacer crawl es de 10 segundos, la directiva de este sitio indica no acceder a las carpetas misc, modules, profiles y themes, sin embargo, si permite acceder a los recursos css, js, a imagenes de estos, permitiendo que los motores de busqueda accedan y rendericen, pero bloqueando la información sensible que no aporta a la indexación de la página. Por ultimo, también bloquea archivos especificos como changelog, cron, e installs, y también restringe el acceso a rutas que corresponden a administración o personal para que estas no sean indexadas, por lo que la página en general restringe el acceso a áreas sensibles e irrelevantes, pero permite el rastreo de recursos como archivos css, js e imágenes.



User-agent: *
Crawl-delay: 10
CSS, JS, Images
Allow: /misc/*.css$
Allow: /misc/*.css?
Allow: /misc/*.js$
Allow: /misc/*.js?
Allow: /misc/*.gif
Allow: /misc/*.jpg
Allow: /misc/*.jpeg
Allow: /misc/*.png
Allow: /modules/*.css$
Allow: /modules/*.css?
Allow: /modules/*.js$
Allow: /modules/*.js?
Allow: /modules/*.gif
Allow: /modules/*.jpg
Allow: /modules/*.jpeg
Allow: /modules/*.png
Allow: /profiles/*.css$
Allow: /profiles/*.css?
Allow: /profiles/*.js$
Allow: /profiles/*.js?
Allow: /profiles/*.gif
Allow: /profiles/*.jpg
Allow: /profiles/*.jpeg
Allow: /profiles/*.png
Allow: /themes/*.css$
Allow: /themes/*.css?
Allow: /themes/*.js$
Allow: /themes/*.js?
Allow: /themes/*.gif
Allow: /themes/*.jpg
Allow: /themes/*.jpeg
Allow: /themes/*.png
Directories
Disallow: /includes/
Disallow: /misc/
Disallow: /modules/
Disallow: /profiles/
Disallow: /scripts/
Disallow: /themes/
Files
Disallow: /CHANGELOG.txt
Disallow: /cron.php
Disallow: /INSTALL.mysql.txt
Disallow: /INSTALL.pgsql.txt
Disallow: /INSTALL.sqlite.txt
Disallow: /install.php
Disallow: /INSTALL.txt
Disallow: /LICENSE.txt
Disallow: /MAINTAINERS.txt
Disallow: /update.php
Disallow: /UPGRADE.txt
Disallow: /xmlrpc.php
Paths (clean URLs)
Disallow: /admin/
Disallow: /comment/reply/
Disallow: /filter/tips/
Disallow: /node/add/
Disallow: /search/
Disallow: /user/register/
Disallow: /user/password/
Disallow: /user/login/
Disallow: /user/logout/
Paths (no clean URLs)
Disallow: /?q=admin/
Disallow: /?q=comment/reply/
Disallow: /?q=filter/tips/
Disallow: /?q=node/add/
Disallow: /?q=search/
Disallow: /?q=user/password/
Disallow: /?q=user/register/
Disallow: /?q=user/login/
Disallow: /?q=user/logout/

https://github.com/robots.txt

Este sitio comparte en su robots un canal de contacto para personas interesadas en hacer crawling, y además de esto también comparte la URL para el API para el mismo fin, las directivas del crawling son las mismas para todos los agentes, sin embargo agrega una condición para el bot baidu, en la cual este tiene que usar un delay de 1 segundo, para el resto de agentes hay restricciones de rutas especificas que parecen corresponder a funcionalidades internas o para evitar cargas innecesarias, la unica ruta con allow espécifico es /*?tab=achievements&achievement=*, por ultimo, también restringe el acceso a account-login y a explodingstuff


If you would like to crawl GitHub contact us via https://support.github.com?tags=dotcom-robots

We also provide an extensive API: https://docs.github.com

User-agent: baidu

crawl-delay: 1


User-agent: *

Disallow: /*/*/pulse
Disallow: /*/*/projects
Disallow: /*/*/forks
Disallow: /*/*/issues/new
Disallow: /*/*/issues/search
Disallow: /*/*/commits/
Disallow: /*/*/branches
Disallow: /*/*/contributors
Disallow: /*/*/tags
Disallow: /*/*/stargazers
Disallow: /*/*/watchers
Disallow: /*/*/network
Disallow: /*/*/graphs
Disallow: /*/*/compare

Disallow: /*/tree/
Disallow: /gist/
Disallow: /*/download
Disallow: /*/revisions
Disallow: /*/commits/*?author
Disallow: /*/commits/*?path
Disallow: /*/comments
Disallow: /*/archive/
Disallow: /*/blame/
Disallow: /*/raw/
Disallow: /*/cache/
Disallow: /.git/
Disallow: */.git/
Disallow: /*.git$
Disallow: /search/advanced
Disallow: /search$
Disallow: /*q=
Disallow: /*.atom$

Disallow: /ekansa/Open-Context-Data
Disallow: /ekansa/opencontext-*
Disallow: */tarball/
Disallow: */zipball/

Disallow: /*source=*
Disallow: /*ref_cta=*
Disallow: /*plan=*
Disallow: /*return_to=*
Disallow: /*ref_loc=*
Disallow: /*setup_organization=*
Disallow: /*source_repo=*
Disallow: /*ref_page=*
Disallow: /*source=*
Disallow: /*referrer=*
Disallow: /*report=*
Disallow: /*author=*
Disallow: /*since=*
Disallow: /*until=*
Disallow: /*commits?author=*
Disallow: /*report-abuse?report=*
Disallow: /*tab=*
Allow: /*?tab=achievements&achievement=*

Disallow: /account-login
Disallow: /Explodingstuff/


### Práctica 2

Con los sitios anteriores (y algún otro que tenga interés en revisar) y usando la biblioteca
urllib.robotparser1 realizar lo siguiente:

Abrir el archivo y establecer el parser, enseguida, con el método .can_fetch() determinar si su User Agent puede examinar
el archivo, y luego, iterar sobre las diferentes entradas y reglas del archivo

In [5]:
import urllib.robotparser
import ssl
import urllib.request

# Ignorar certificado SSL
ssl_context = ssl._create_unverified_context()

websites = [
    "https://www.cicese.mx/robots.txt",
    "https://www.uabc.mx/robots.txt",
    "https://www.unam.mx/robots.txt",
    "https://www.github.com/robots.txt",
    "https://www.halopedia.org/robots.txt"
]

# Función para revisar con can_fetch
def can_fetch(user_agent, url):
    rp = urllib.robotparser.RobotFileParser()
    rp.set_url(url)
    rp.read()
    return rp.can_fetch(user_agent, url)

# Función para iterar las reglas
def iterate_rules(robots_content):
    rfp = urllib.robotparser.RobotFileParser()
    rfp.parse(robots_content.splitlines())
    entries = [rfp.default_entry, *rfp.entries] if rfp.default_entry else rfp.entries

    for entry in entries:
        for ruleline in entry.rulelines:
            yield (entry.useragents, ruleline.path, ruleline.allowance)

results = {}

for site in websites:
    try:
        with urllib.request.urlopen(site, context=ssl_context) as response:
            robots_content = response.read().decode('utf-8')
            rules = list(iterate_rules(robots_content))
            results[site] = (robots_content, rules)  # Save both content and rules
    except Exception as e:
        print(f"Skipping {site} due to error: {e}")
        continue

# Desplegar resultados
for site, (robots_content, rules) in results.items():
    print(f"Rules for {site}:")
    if not rules:
        print("  No rules found. Here's the original robots.txt content:")
        print(robots_content.strip() or "  (robots.txt is empty)")
    else:
        for rule in rules:
            useragents, path, allowance = rule
            print(f"  User Agents: {', '.join(useragents) if useragents else '*'}")
            print(f"  Path: {path}")
            print(f"  Allowed: {'Yes' if allowance else 'No'}")
    print("\n")




Skipping https://www.unam.mx/robots.txt due to error: HTTP Error 403: Forbidden
Rules for https://www.cicese.mx/robots.txt:
  User Agents: SemrushBot
  Path: /
  Allowed: No
  User Agents: SemrushBot-SA
  Path: /
  Allowed: No
  User Agents: SemrushBot-BM
  Path: /
  Allowed: No
  User Agents: SemrushBot-CT
  Path: /
  Allowed: No
  User Agents: SemrushBot-SWA
  Path: /
  Allowed: No
  User Agents: SemrushBot-SI
  Path: /
  Allowed: No
  User Agents: SemrushBot-BA
  Path: /
  Allowed: No
  User Agents: BLEXBot
  Path: /
  Allowed: No


Rules for https://www.uabc.mx/robots.txt:
  User Agents: *
  Path: /wp-admin/
  Allowed: No
  User Agents: *
  Path: /wp-admin/admin-ajax.php
  Allowed: Yes
  User Agents: *
  Path: /wp-content/uploads/wpforms/
  Allowed: No


Rules for https://www.github.com/robots.txt:
  No rules found. Here's the original robots.txt content:
# If you would like to crawl GitHub contact us via https://support.github.com?tags=dotcom-robots
# We also provide an extensive 

Como se puede ver en la consola, tres de los sitios visitados tienen reglas especificas para el agente, un sitio no tienen reglas y un sitio da error 403, lo cual indica que el servidor está bloqueando mi solicitud, después de intentar con distintos agentes llegué a lo mismo con este sitio.

In [8]:
import re
import urllib.robotparser
import ssl
import urllib.request

ssl_context = ssl._create_unverified_context()

websites = [
    "https://www.cicese.mx/robots.txt",
    "https://www.uabc.mx/robots.txt",
    "https://www.unam.mx/robots.txt",
    "https://www.github.com/robots.txt",
    "https://www.halopedia.org/robots.txt"
]

# Función para crear patrones regex a partir de las rutas de robots.txt
# Esto permite manejar wildcards (*) y el símbolo de fin de línea ($)
def get_robots_pattern(path):
    ending = '.*'  # Por defecto, permitimos cualquier cosa después de la ruta
    if path.endswith('$'):
        path = path[:-1]  # Eliminar $ 
        ending = '$'
    parts = path.split('*')  # Dividir la ruta donde haya '*'
    parts = map(urllib.parse.quote, map(re.escape, parts))  
    return '.*'.join(parts) + ending  # Reunir los fragmentos en un patrón regex

# Parche para la clase RuleLine de urllib.robotparser qque permite wildcards
def _rule_line_init__(self, path, allowance):
    if path == '' and not allowance:
        allowance = True
    path = urllib.parse.unquote(urllib.parse.unquote(path))  # Decodificar URL
    self.pattern = re.compile(get_robots_pattern(path)) 
    self.path = path  
    self.allowance = allowance  # Guardar si se permite (True) o no (False)

# Verificar si la regla aplica a una URL específica
# Devuelve True si la URL coincide con el patrón de la regla
def _rule_line_applies_to(self, filename):
    return True if self.pattern.match(filename) else False

# Aplicar el parche a la clase RuleLine
urllib.robotparser.RuleLine.__init__ = _rule_line_init__
urllib.robotparser.RuleLine.applies_to = _rule_line_applies_to

# Función para iterar sobre las reglas
def iterate_rules(robots_content):
    rfp = urllib.robotparser.RobotFileParser()
    rfp.parse(robots_content.splitlines())  # Analizar el contenido línea por línea
    entries = [rfp.default_entry, *rfp.entries] if rfp.default_entry else rfp.entries

    for entry in entries:
        for ruleline in entry.rulelines:
            yield (entry.useragents, ruleline.path, ruleline.allowance)  # Devolver User-Agent, ruta y permiso

results = {}

for site in websites:
    try:
        with urllib.request.urlopen(site, context=ssl_context) as response:
            robots_content = response.read().decode('utf-8')  # Leer y decodificar el contenido
            rules = list(iterate_rules(robots_content))  # Obtener las reglas
            results[site] = (robots_content, rules)  # Guardar tanto el contenido como las reglas
            
    except Exception as e:
        print(f"Skipping {site} due to error: {e}")  # Saltar el sitio en caso de error
        continue

# Mostrar los resultados
for site, (robots_content, rules) in results.items():
    print(f"Rules for {site}:")
    if not rules:
        print("  No rules found. Here's the original robots.txt content:")
        print(robots_content.strip() or "  (robots.txt is empty)")  # Imprimir el contenido si no hay reglas
    else:
        for rule in rules:
            useragents, path, allowance = rule
            print(f"  User Agents: {', '.join(useragents) if useragents else '*'}")  # Imprimir User-Agent
            print(f"  Path: {path}")  # Imprimir la ruta
            print(f"  Allowed: {'Yes' if allowance else 'No'}")  # Indicar si está permitido o no
    print("\n")


Skipping https://www.unam.mx/robots.txt due to error: HTTP Error 403: Forbidden
Rules for https://www.cicese.mx/robots.txt:
  User Agents: SemrushBot
  Path: /
  Allowed: No
  User Agents: SemrushBot-SA
  Path: /
  Allowed: No
  User Agents: SemrushBot-BM
  Path: /
  Allowed: No
  User Agents: SemrushBot-CT
  Path: /
  Allowed: No
  User Agents: SemrushBot-SWA
  Path: /
  Allowed: No
  User Agents: SemrushBot-SI
  Path: /
  Allowed: No
  User Agents: SemrushBot-BA
  Path: /
  Allowed: No
  User Agents: BLEXBot
  Path: /
  Allowed: No


Rules for https://www.uabc.mx/robots.txt:
  User Agents: *
  Path: /wp-admin/
  Allowed: No
  User Agents: *
  Path: /wp-admin/admin-ajax.php
  Allowed: Yes
  User Agents: *
  Path: /wp-content/uploads/wpforms/
  Allowed: No


Rules for https://www.github.com/robots.txt:
  No rules found. Here's the original robots.txt content:
# If you would like to crawl GitHub contact us via https://support.github.com?tags=dotcom-robots
# We also provide an extensive 

Para esta parte del código decidí agregar más detalles en los comentarios ya que el procesado lo saqué del ejemplo y no tengo mucho que agregar de mi parte más que el entendimiento del proceso

## Práctica 3

Algunos sitios también ofrecen un mapa del sitio, por lo que de ser así, conviene
analizarlo. Entonces, con los sitios que revisó anteriormente, verifique mediante su
browser si existe el archivo sitemap.xml (o bien revise en robots.txt si se indica
alguno) y comente sobre lo que encuentre.

In [ ]:
%pip install advertools

In [11]:
import advertools as adv  # Para analizar los sitemaps

def extract_sitemap(robots_content):
    sitemaps = re.findall(r'Sitemap:\s*(\S+)', robots_content)
    return sitemaps

for site in websites:
    try:
        with urllib.request.urlopen(site, context=ssl_context) as response:
            sitemaps = extract_sitemap(robots_content)
    except Exception as e:
        print(f"Skipping {site} due to error: {e}")
        continue

    # Procesar y mostrar el sitemap si existe
    if sitemaps:
        print("  Sitemaps found:")
        for sitemap_url in sitemaps:
            print(f"    {sitemap_url}")
            try:
                sitemap_df = adv.sitemap_to_df(sitemap_url)
                print(sitemap_df[['loc', 'lastmod', 'changefreq', 'priority']].head())
            except Exception as e:
                print(f"    Could not process sitemap: {e}")
    else:
        print("  No sitemap found.")

    print("\n")


  Sitemaps found:
    https://www.halopedia.org/sitemap/sitemap-index-wiki.xml


2025-02-03 16:35:35,670 | INFO | sitemaps.py:623 | sitemap_to_df | Getting https://www.halopedia.org/sitemap/sitemap-wiki-NS_12-0.xml.gz
2025-02-03 16:35:35,672 | INFO | sitemaps.py:623 | sitemap_to_df | Getting https://www.halopedia.org/sitemap/sitemap-wiki-NS_100-0.xml.gz
2025-02-03 16:35:35,673 | INFO | sitemaps.py:623 | sitemap_to_df | Getting https://www.halopedia.org/sitemap/sitemap-wiki-NS_104-0.xml.gz
2025-02-03 16:35:35,700 | INFO | sitemaps.py:623 | sitemap_to_df | Getting https://www.halopedia.org/sitemap/sitemap-wiki-NS_4-0.xml.gz
2025-02-03 16:35:35,829 | INFO | sitemaps.py:623 | sitemap_to_df | Getting https://www.halopedia.org/sitemap/sitemap-wiki-NS_6-1.xml.gz
2025-02-03 16:35:35,867 | INFO | sitemaps.py:623 | sitemap_to_df | Getting https://www.halopedia.org/sitemap/sitemap-wiki-NS_14-0.xml.gz
2025-02-03 16:35:35,927 | INFO | sitemaps.py:623 | sitemap_to_df | Getting https://www.halopedia.org/sitemap/sitemap-wiki-NS_102-0.xml.gz
2025-02-03 16:35:36,025 | INFO | sitemap

    Could not process sitemap: "['changefreq'] not in index"


  Sitemaps found:
    https://www.halopedia.org/sitemap/sitemap-index-wiki.xml


2025-02-03 16:35:58,504 | INFO | sitemaps.py:623 | sitemap_to_df | Getting https://www.halopedia.org/sitemap/sitemap-wiki-NS_4-0.xml.gz
2025-02-03 16:35:58,507 | INFO | sitemaps.py:623 | sitemap_to_df | Getting https://www.halopedia.org/sitemap/sitemap-wiki-NS_104-0.xml.gz
2025-02-03 16:35:58,517 | INFO | sitemaps.py:623 | sitemap_to_df | Getting https://www.halopedia.org/sitemap/sitemap-wiki-NS_12-0.xml.gz
2025-02-03 16:35:58,518 | INFO | sitemaps.py:623 | sitemap_to_df | Getting https://www.halopedia.org/sitemap/sitemap-wiki-NS_100-0.xml.gz
2025-02-03 16:35:58,667 | INFO | sitemaps.py:623 | sitemap_to_df | Getting https://www.halopedia.org/sitemap/sitemap-wiki-NS_6-1.xml.gz
2025-02-03 16:35:58,709 | INFO | sitemaps.py:623 | sitemap_to_df | Getting https://www.halopedia.org/sitemap/sitemap-wiki-NS_14-0.xml.gz
2025-02-03 16:35:58,767 | INFO | sitemaps.py:623 | sitemap_to_df | Getting https://www.halopedia.org/sitemap/sitemap-wiki-NS_102-0.xml.gz
2025-02-03 16:35:58,858 | INFO | sitemap

    Could not process sitemap: "['changefreq'] not in index"


Skipping https://www.unam.mx/robots.txt due to error: HTTP Error 403: Forbidden
  Sitemaps found:
    https://www.halopedia.org/sitemap/sitemap-index-wiki.xml


2025-02-03 16:36:00,572 | INFO | sitemaps.py:623 | sitemap_to_df | Getting https://www.halopedia.org/sitemap/sitemap-wiki-NS_12-0.xml.gz
2025-02-03 16:36:00,575 | INFO | sitemaps.py:623 | sitemap_to_df | Getting https://www.halopedia.org/sitemap/sitemap-wiki-NS_104-0.xml.gz
2025-02-03 16:36:00,665 | INFO | sitemaps.py:623 | sitemap_to_df | Getting https://www.halopedia.org/sitemap/sitemap-wiki-NS_100-0.xml.gz
2025-02-03 16:36:00,667 | INFO | sitemaps.py:623 | sitemap_to_df | Getting https://www.halopedia.org/sitemap/sitemap-wiki-NS_4-0.xml.gz
2025-02-03 16:36:00,746 | INFO | sitemaps.py:623 | sitemap_to_df | Getting https://www.halopedia.org/sitemap/sitemap-wiki-NS_14-0.xml.gz
2025-02-03 16:36:00,791 | INFO | sitemaps.py:623 | sitemap_to_df | Getting https://www.halopedia.org/sitemap/sitemap-wiki-NS_6-1.xml.gz
2025-02-03 16:36:00,957 | INFO | sitemaps.py:623 | sitemap_to_df | Getting https://www.halopedia.org/sitemap/sitemap-wiki-NS_0-0.xml.gz
2025-02-03 16:36:00,961 | INFO | sitemaps.

    Could not process sitemap: "['changefreq'] not in index"


  Sitemaps found:
    https://www.halopedia.org/sitemap/sitemap-index-wiki.xml


2025-02-03 16:36:02,545 | INFO | sitemaps.py:623 | sitemap_to_df | Getting https://www.halopedia.org/sitemap/sitemap-wiki-NS_104-0.xml.gz
2025-02-03 16:36:02,548 | INFO | sitemaps.py:623 | sitemap_to_df | Getting https://www.halopedia.org/sitemap/sitemap-wiki-NS_12-0.xml.gz
2025-02-03 16:36:02,549 | INFO | sitemaps.py:623 | sitemap_to_df | Getting https://www.halopedia.org/sitemap/sitemap-wiki-NS_100-0.xml.gz
2025-02-03 16:36:02,550 | INFO | sitemaps.py:623 | sitemap_to_df | Getting https://www.halopedia.org/sitemap/sitemap-wiki-NS_4-0.xml.gz
2025-02-03 16:36:02,705 | INFO | sitemaps.py:623 | sitemap_to_df | Getting https://www.halopedia.org/sitemap/sitemap-wiki-NS_14-0.xml.gz
2025-02-03 16:36:02,738 | INFO | sitemaps.py:623 | sitemap_to_df | Getting https://www.halopedia.org/sitemap/sitemap-wiki-NS_6-1.xml.gz
2025-02-03 16:36:02,797 | INFO | sitemaps.py:623 | sitemap_to_df | Getting https://www.halopedia.org/sitemap/sitemap-wiki-NS_102-0.xml.gz
2025-02-03 16:36:02,903 | INFO | sitemap

    Could not process sitemap: "['changefreq'] not in index"




Para el procesado del sitemap se usó advertools, esta herramienta cuenta con una buena documentación al respecto de robots y sitemaps en https://advertools.readthedocs.io/en/master/advertools.sitemaps.html, a pesar de ser solo el final de la práctica la documentación es interesante y tiene buenos ejemplos. Respecto a la página de unam, intenté con otros agentes, sin embargo me seguía bloqueando y las alternativas que se me ocurrieron para poder acceder se desviaban mucho de la práctica, por lo que para el reporte lo dejé así 